File: camerax_android/lib/src/camera2/capture_request_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/camera2.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class CaptureRequest$KeyImpl<T> implements CaptureRequest$Key<T> {
  final CaptureRequestKeyProxyApi api;

  CaptureRequest$KeyImpl.internal(this.api);

  @override
  Future<String> getName() => api.getName();
}

final class CaptureRequestImpl extends CameraMetadataImpl<CaptureRequest$Key>
    implements CaptureRequest {
  final CaptureRequestProxyApi api;

  CaptureRequestImpl.internal(this.api);

  @override
  Future<T?> get<T>(CaptureRequest$Key<T> key) {
    final keyApi = key.api;
    if (keyApi is CaptureRequestIntKeyProxyApi) {
      return api.getInt(keyApi).then((e) => e as T?);
    }
    if (keyApi is CaptureRequestLongKeyProxyApi) {
      return api.getLong(keyApi).then((e) => e as T?);
    }
    if (keyApi is CaptureRequestFloatKeyProxyApi) {
      return api.getFloat(keyApi).then((e) => e as T?);
    }
    if (keyApi is CaptureRequestBooleanKeyProxyApi) {
      return api.getBoolean(keyApi).then((e) => e as T?);
    }
    if (keyApi is CaptureRequestCameraMetadataControlModeKeyProxyApi) {
      return api
          .getCameraMetadataControlMode(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi is CaptureRequestCameraMetadataControlAeModeKeyProxyApi) {
      return api
          .getCameraMetadataControlAeMode(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi is CaptureRequestCameraMetadataControlAfModeKeyProxyApi) {
      return api
          .getCameraMetadataControlAfMode(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi is CaptureRequestCameraMetadataControlAwbModeKeyProxyApi) {
      return api
          .getCameraMetadataControlAwbMode(keyApi)
          .then((e) => e?.impl as T?);
    }
    throw TypeError();
  }

  @override
  Future<bool> isReprocess() => api.isReprocess();
}

final class CaptureRequestChannelImpl extends CaptureRequestChannel {
  @override
  CaptureRequest$Key<bool> get controlAeLock =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlAeLock);

  @override
  CaptureRequest$Key<CameraMetadata$ControlAeMode> get controlAeMode =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlAeMode);

  @override
  CaptureRequest$Key<CameraMetadata$ControlAfMode> get controlAfMode =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlAfMode);

  @override
  CaptureRequest$Key<bool> get controlAwbLock =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlAwbLock);

  @override
  CaptureRequest$Key<CameraMetadata$ControlAwbMode> get controlAwbMode =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlAwbMode);

  @override
  CaptureRequest$Key<CameraMetadata$ControlMode> get controlMode =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.controlMode);

  @override
  CaptureRequest$Key<double> get lensAperture =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.lensAperture);

  @override
  CaptureRequest$Key<int> get sensorExposureTime =>
      CaptureRequest$KeyImpl.internal(
        CaptureRequestProxyApi.sensorExposureTime,
      );

  @override
  CaptureRequest$Key<int> get sensorSensitivity =>
      CaptureRequest$KeyImpl.internal(CaptureRequestProxyApi.sensorSensitivity);
}

extension CaptureRequest$KeyX on CaptureRequest$Key {
  CaptureRequestKeyProxyApi get api {
    final impl = this;
    if (impl is! CaptureRequest$KeyImpl) throw TypeError();
    return impl.api;
  }
}

extension CaptureRequestProxyApiX on CaptureRequestProxyApi {
  CaptureRequest get impl => CaptureRequestImpl.internal(this);
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/camera2/CaptureRequestImpl.kt
package dev.zeekr.camerax_android.camera2

import android.hardware.camera2.CaptureRequest
import dev.zeekr.camerax_android.CameraMetadataControlAeModeApi
import dev.zeekr.camerax_android.CameraMetadataControlAfModeApi
import dev.zeekr.camerax_android.CameraMetadataControlAwbModeApi
import dev.zeekr.camerax_android.CameraMetadataControlModeApi
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiCaptureRequestKeyProxyApi
import dev.zeekr.camerax_android.PigeonApiCaptureRequestProxyApi

class CaptureRequestImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiCaptureRequestProxyApi(registrar) {
    class KeyImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiCaptureRequestKeyProxyApi(registrar) {
        override fun getName(pigeon_instance: Key): String {
            return pigeon_instance.instance.name
        }
    }

    open class Key(open val instance: CaptureRequest.Key<*>)
    class IntKey(override val instance: CaptureRequest.Key<Int>) : Key(instance)
    class LongKey(override val instance: CaptureRequest.Key<Long>) : Key(instance)
    class FloatKey(override val instance: CaptureRequest.Key<Float>) : Key(instance)
    class BooleanKey(override val instance: CaptureRequest.Key<Boolean>) : Key(instance)
    class CameraMetadataControlModeKey(override val instance: CaptureRequest.Key<Int>) : Key(instance)
    class CameraMetadataControlAeModeKey(override val instance: CaptureRequest.Key<Int>) : Key(instance)
    class CameraMetadataControlAfModeKey(override val instance: CaptureRequest.Key<Int>) : Key(instance)
    class CameraMetadataControlAwbModeKey(override val instance: CaptureRequest.Key<Int>) : Key(instance)

    override fun controlAeLock(): BooleanKey {
        return BooleanKey(CaptureRequest.CONTROL_AE_LOCK)
    }

    override fun controlAeMode(): CameraMetadataControlAeModeKey {
        return CameraMetadataControlAeModeKey(CaptureRequest.CONTROL_AE_MODE)
    }

    override fun controlAfMode(): CameraMetadataControlAfModeKey {
        return CameraMetadataControlAfModeKey(CaptureRequest.CONTROL_AF_MODE)
    }

    override fun controlAwbLock(): BooleanKey {
        return BooleanKey(CaptureRequest.CONTROL_AWB_LOCK)
    }

    override fun controlAwbMode(): CameraMetadataControlAwbModeKey {
        return CameraMetadataControlAwbModeKey(CaptureRequest.CONTROL_AWB_MODE)
    }

    override fun controlMode(): CameraMetadataControlModeKey {
        return CameraMetadataControlModeKey(CaptureRequest.CONTROL_MODE)
    }

    override fun lensAperture(): FloatKey {
        return FloatKey(CaptureRequest.LENS_APERTURE)
    }

    override fun sensorExposureTime(): LongKey {
        return LongKey(CaptureRequest.SENSOR_EXPOSURE_TIME)
    }

    override fun sensorSensitivity(): IntKey {
        return IntKey(CaptureRequest.SENSOR_SENSITIVITY)
    }

    override fun getInt(pigeon_instance: CaptureRequest, key: IntKey): Long? {
        return pigeon_instance.get(key.instance)?.toLong()
    }

    override fun getLong(pigeon_instance: CaptureRequest, key: LongKey): Long? {
        return pigeon_instance.get(key.instance)
    }

    override fun getFloat(pigeon_instance: CaptureRequest, key: FloatKey): Double? {
        return pigeon_instance.get(key.instance)?.toDouble()
    }

    override fun getBoolean(pigeon_instance: CaptureRequest, key: BooleanKey): Boolean? {
        return pigeon_instance.get(key.instance)
    }

    override fun getCameraMetadataControlMode(
        pigeon_instance: CaptureRequest, key: CameraMetadataControlModeKey
    ): CameraMetadataControlModeApi? {
        return pigeon_instance.get(key.instance)?.cameraMetadataControlModeApi
    }

    override fun getCameraMetadataControlAeMode(
        pigeon_instance: CaptureRequest, key: CameraMetadataControlAeModeKey
    ): CameraMetadataControlAeModeApi? {
        return pigeon_instance.get(key.instance)?.cameraMetadataControlAeModeApi
    }

    override fun getCameraMetadataControlAfMode(
        pigeon_instance: CaptureRequest, key: CameraMetadataControlAfModeKey
    ): CameraMetadataControlAfModeApi? {
        return pigeon_instance.get(key.instance)?.cameraMetadataControlAfModeApi
    }

    override fun getCameraMetadataControlAwbMode(
        pigeon_instance: CaptureRequest, key: CameraMetadataControlAwbModeKey
    ): CameraMetadataControlAwbModeApi? {
        return pigeon_instance.get(key.instance)?.cameraMetadataControlAwbModeApi
    }

    override fun isReprocess(pigeon_instance: CaptureRequest): Boolean {
        return pigeon_instance.isReprocess
    }
}